src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultProjectConfigurationFilePathStore.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(ProjectConfigurationFilePathStore))]
+    internal class DefaultProjectConfigurationFilePathStore : ProjectConfigurationFilePathStore
+    {
+        private readonly Dictionary<string, string> _mappings;
+        private readonly object _mappingsLock;
+
+        public override event EventHandler<ProjectConfigurationFilePathChangedEventArgs> Changed;
+
+        [ImportingConstructor]
+        public DefaultProjectConfigurationFilePathStore()
+        {
+            _mappings = new Dictionary<string, string>(FilePathComparer.Instance);
+            _mappingsLock = new object();
+        }
+
+        public override IReadOnlyDictionary<string, string> GetMappings() => new Dictionary<string, string>(_mappings);
+
+        public override void Set(string projectFilePath, string configurationFilePath)
+        {
+            if (projectFilePath is null)
+            {
+                throw new ArgumentNullException(nameof(projectFilePath));
+            }
+
+            if (configurationFilePath is null)
+            {
+                throw new ArgumentNullException(nameof(configurationFilePath));
+            }
+
+            lock (_mappingsLock)
```

Comment:
Purposefully not using a `ConcurrentDictionary` for `_mappings` because of this lock. Basically the need to retrieve the value, potentially no-op and then update the value. Basically making _mappings concurrent wouldn't buy us anything except for the extra overhead given we have to lock anyways.